### ISSUE TITLE
slow (does not finish within one hour)

### ISSUE THREAD

**Issue by JoKalliauer** [scour-project/scour#184]
Prozessing files over 10MB are very slow, and does not finish within several hours.

first example (contains 483.840 lines, 28MB)
https://upload.wikimedia.org/wikipedia/commons/f/ff/1_42_polytope_7-cube.svg
[Rendered by Wikipedia](https://commons.wikimedia.org/wiki/File:1_42_polytope_7-cube.svg)

Second example (70MB)
https://upload.wikimedia.org/wikipedia/commons/archive/9/9d/20171030170157%21E8out96g16.svg
[Rendering by Wikipedia](https://commons.wikimedia.org/wiki/File:E8out96g16.svg) (after optimisation/pathmerging with [svgcleaner](https://github.com/RazrFalcon/svgcleaner) ans [svgo](https://github.com/svg/svgo))

I ran the first file in Feb18 and the second in Oct17 over night wihtout success. (So they were tested with old versions of scour, but also in the actuall version it is more than 1000times slower than [svgcleaner](https://github.com/RazrFalcon/svgcleaner).)

If you want to optimise bulk-processing of big files I recommend to download the svg-files from:
https://commons.wikimedia.org/wiki/Category:Large_SVG_files

To not download them separately you can use Imker that downloads al files in a chosen category, (f.e (Large_SVG_files)[https://commons.wikimedia.org/wiki/Category:Large_SVG_files]) for you
https://commons.wikimedia.org/wiki/Commons:Imker_(batch_download)

**Comment by nthykier** [comment]
For the first one (1_42_polytope_7-cube.svg), I see a runtime of about 8 minutes with python3 on master@6ea126d290e29c990244d9df17f00f84b7c07108.  While far from optimal, it seem like it is now doable to process and we are looking at some constant time improvements.

The 1_42_polytope_7-cube.svg consists almost exclusively on "<line ..>" rules.  Of the 491713 lines in the image, 483840 starts with `<line` (remaining items appear to be a few thousand `<circle>`-entries and a few "<g>` wrapping the other tags).

I have not have time to look at the other image in details yet.

**Comment by nthykier** [comment]
The number of each type of tag (by looking only at the start of each line) in the the 70MB image:

```
     12 <ellipse
     54 <text
     91 <clipPath
     92 <g
     92 </g>
    105 <polyline
 142351 <path
 355157 <polygon
```

(I omitted the items that only appeared once in the above)

Basically, it is half a million polygon and paths (in total).  Have not profiled it (and won't have time to run it either today)

**Comment by nthykier** [comment]
A slight complication: 1_42_polytope_7-cube.svg requires 1.3GB RAM to process with scour (using python3).  With 20171030170157%21E8out96g16.svg being over 2 times as large, it will probably take at least 2 times as much memory.

We probably have to look into memory optimization as well to improve these test cases.

**Comment by nthykier** [comment]
@JoKalliauer , I have now successfully run both of the examples you presented here:

- They took about 7 to 8 minute each on my hardware (single run)
- They required 1.3 to 1.6GB of RAM to process (if you have less than that, it will probably become painful performance-wise).
- The "smaller one" is reduced to 90.7% of the original size and the larger one to 68.4% of the original size (which could explain why there is less difference in the memory usage).
- I tested with master (8ddb7d8913f862006cdc99d82dcb97e4f209f418).
- The above is based on python 3.6 using a Linux machine (CPython, not pypy).

As noted in #186, using python2.7 could add a factor 5-6 on top of that (Using an older version of python 3.X might as well).  Running scour in cygwin may cause additional slow on top of that as it includes an emulation layer - However, I have no clue how much (scour does not seem like it would involve the emulation layer all that much).

In conclusion:

- The current (unreleased) version of scour would do vastly better than what you experienced (at least assuming you can use python 3.6)
- There is still room for improvement on these images.

**Comment by JoKalliauer** [comment]
@nthykier 
I have a one year old computer with 16GB, Quadcore: 4*2.5GHz, SSD, there that should be enough RAM, since I the OS uses about 30% of RAM.

I now run it again the first took ~6min and the second took ~7min. That is great, and therefore I will close the bug.
But before that I want to know how can I derive the used python-version, since I installed Python 2.7 and few days ago Python 3.x 

If I type:
>$ python --version
Python 2.7.14

but if I type:
>$ pip --version
pip 10.0.1 from c:\python36\lib\site-packages\pip (python 3.6)

but scour does not specify the python-version
>$ scour --version
0.36
